packages/bierzo-wallet/src/logic/connection.ts

```diff
@@ -0,0 +1,56 @@
+import { BlockchainConnection } from '@iov/bcp';
+import { BnsConnection } from '@iov/bns';
+import { EthereumConnection } from '@iov/ethereum';
+import { LiskConnection } from '@iov/lisk';
+import { singleton } from 'medulas-react-components/lib/utils/singleton';
+import { ChainSpec, loadConfig } from '../utils/config';
+
+async function createEthereumConnection(url: string): Promise<BlockchainConnection> {
+  return EthereumConnection.establish(url, {});
+}
+
+let getEthereumConnection = singleton<typeof createEthereumConnection>(createEthereumConnection);
+
+async function createBnsConnection(url: string): Promise<BlockchainConnection> {
+  return BnsConnection.establish(url);
+}
+
+let getBnsConnection = singleton<typeof createBnsConnection>(createBnsConnection);
+
+async function createLiskConnection(url: string): Promise<BlockchainConnection> {
+  return LiskConnection.establish(url);
+}
+
+let getLiskConnection = singleton<typeof createLiskConnection>(createLiskConnection);
+
+export async function getConnectionFor(spec: ChainSpec): Promise<BlockchainConnection> {
+  const url = spec.bootstrapNodes[0];
+  if (spec.codecType === 'eth') {
+    return getEthereumConnection(url);
+  }
+
+  if (spec.codecType === 'bns') {
+    return getBnsConnection(url);
+  }
+
+  if (spec.codecType === 'lsk') {
+    return getLiskConnection(url);
+  }
+
+  throw new Error('Chain spec not supported');
+}
+
+export async function disconnect(): Promise<void> {
+  const config = await loadConfig();
```

Comment:
@webmaster128 where in the code we should call this? As far as I understand we rely on WS connection, and if the browser is closed, automatically the connection is cleared.

Is this related to some kind of possible logout action?

Reply:
Right. At the moment you don't need to call it at all. If we lazy load the connections and keep them alive forever, we are good. Maybe we have a good reason to disconnect and reconnect later on, but at the moment I don't see one.

The problem with the previous code was that every call of `getTokens()` created new connections which were not disconnected, leading to 3, 6, 9, 12, … open connections. This issue does not exist anymore.